File: java/exercises/practice/unit_test_bfd52424_a953_408f_bd78_191d05dc0837/task_scheduler/src/test/java/TaskSchedulerTest.java
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import java.util.*;

public class TaskSchedulerTest {

    @Test
    public void testEmptyTasksAndMachines() {
        List<Task> tasks = new ArrayList<>();
        List<Machine> machines = new ArrayList<>();

        TaskScheduler scheduler = new TaskScheduler();
        int makespan = scheduler.schedule(tasks, machines);
        assertEquals(0, makespan, "No tasks and no machines should result in a makespan of 0.");
    }

    @Test
    public void testEmptyTasksNonEmptyMachines() {
        List<Task> tasks = new ArrayList<>();
        List<Machine> machines = Arrays.asList(
            new Machine("M1", 4, 16),
            new Machine("M2", 8, 32)
        );

        TaskScheduler scheduler = new TaskScheduler();
        int makespan = scheduler.schedule(tasks, machines);
        assertEquals(0, makespan, "No tasks should result in a makespan of 0.");
    }

    @Test
    public void testSingleTaskNoDependencies() {
        Task task = new Task("T1", 2, 4, 10, new HashSet<>());
        List<Task> tasks = Arrays.asList(task);
        List<Machine> machines = Arrays.asList(new Machine("M1", 4, 16));

        TaskScheduler scheduler = new TaskScheduler();
        int makespan = scheduler.schedule(tasks, machines);
        assertEquals(10, makespan, "A single task should complete in its execution time.");
    }

    @Test
    public void testMultipleTasksNoDependencies() {
        // Three independent tasks that can run concurrently given ample resources.
        Task task1 = new Task("T1", 2, 4, 10, new HashSet<>());
        Task task2 = new Task("T2", 2, 4, 15, new HashSet<>());
        Task task3 = new Task("T3", 2, 4, 5, new HashSet<>());
        List<Task> tasks = Arrays.asList(task1, task2, task3);
        List<Machine> machines = Arrays.asList(
            new Machine("M1", 4, 16),
            new Machine("M2", 4, 16)
        );

        TaskScheduler scheduler = new TaskScheduler();
        int makespan = scheduler.schedule(tasks, machines);
        // Expected makespan is the maximum execution time (15) if tasks are run concurrently.
        assertEquals(15, makespan, "Concurrent tasks should result in the maximum individual execution time as the makespan.");
    }

    @Test
    public void testTasksWithDependencies() {
        // Chain dependencies: T1 -> T2 -> T3 (T2 depends on T1, T3 depends on T2)
        Task task1 = new Task("T1", 2, 4, 10, new HashSet<>());
        Task task2 = new Task("T2", 2, 4, 15, new HashSet<>(Arrays.asList("T1")));
        Task task3 = new Task("T3", 2, 4, 5, new HashSet<>(Arrays.asList("T2")));
        List<Task> tasks = Arrays.asList(task1, task2, task3);
        List<Machine> machines = Arrays.asList(new Machine("M1", 4, 16));

        TaskScheduler scheduler = new TaskScheduler();
        int makespan = scheduler.schedule(tasks, machines);
        // Sequential execution due to dependencies: total time = 10 + 15 + 5 = 30.
        assertEquals(30, makespan, "Dependent tasks must run sequentially resulting in the summed execution time.");
    }

    @Test
    public void testMultipleSchedulingOptions() {
        // T1 and T2 can run concurrently, and T3 depends on both T1 and T2.
        Task task1 = new Task("T1", 2, 4, 10, new HashSet<>());
        Task task2 = new Task("T2", 2, 4, 12, new HashSet<>());
        Task task3 = new Task("T3", 2, 4, 8, new HashSet<>(Arrays.asList("T1", "T2")));
        List<Task> tasks = Arrays.asList(task1, task2, task3);
        List<Machine> machines = Arrays.asList(
            new Machine("M1", 4, 16),
            new Machine("M2", 4, 16)
        );

        TaskScheduler scheduler = new TaskScheduler();
        int makespan = scheduler.schedule(tasks, machines);
        // T1 and T2 run concurrently, so the max time is 12, then T3 takes 8, total = 20.
        assertEquals(20, makespan, "Tasks with partial parallelism should yield an optimized makespan.");
    }

    @Test
    public void testCycleDetection() {
        // Create cyclic dependency: T1 -> T2 and T2 -> T1.
        Task task1 = new Task("T1", 2, 4, 10, new HashSet<>(Arrays.asList("T2")));
        Task task2 = new Task("T2", 2, 4, 15, new HashSet<>(Arrays.asList("T1")));
        List<Task> tasks = Arrays.asList(task1, task2);
        List<Machine> machines = Arrays.asList(new Machine("M1", 4, 16));

        TaskScheduler scheduler = new TaskScheduler();
        Exception exception = assertThrows(IllegalArgumentException.class, () -> {
            scheduler.schedule(tasks, machines);
        });
        String expectedMessage = "cyclic dependency";
        String actualMessage = exception.getMessage();
        assertTrue(actualMessage.toLowerCase().contains(expectedMessage), "A cyclic dependency should be detected.");
    }

    @Test
    public void testInsufficientResources() {
        // Task that requires more CPU cores than any available machine can provide.
        Task task = new Task("T1", 10, 4, 10, new HashSet<>());
        List<Task> tasks = Arrays.asList(task);
        List<Machine> machines = Arrays.asList(
            new Machine("M1", 4, 16),
            new Machine("M2", 4, 16)
        );

        TaskScheduler scheduler = new TaskScheduler();
        Exception exception = assertThrows(IllegalArgumentException.class, () -> {
            scheduler.schedule(tasks, machines);
        });
        String expectedMessage = "insufficient resources";
        String actualMessage = exception.getMessage();
        assertTrue(actualMessage.toLowerCase().contains(expectedMessage), "The scheduler should flag tasks with resource requirements exceeding machine capacity.");
    }
}

File: cpp/exercises/practice/unit_test_0fa8d1be_fed7_49da_b804_2d01aea31f24/task_scheduler/task_scheduler_test.cpp
#include "catch.hpp"
#include "task_scheduler.h"
#include <vector>

TEST_CASE("Basic sequential tasks") {
    std::vector<int> duration = {3, 2, 4};
    std::vector<int> resource_requirement = {2, 3, 2};
    std::vector<std::vector<int>> dependencies = {{}, {0}, {1}};
    int resource_limit = 4;

    auto result = task_scheduler::schedule(3, duration, resource_requirement, dependencies, resource_limit);
    
    REQUIRE(result.size() == 3);
    REQUIRE(result[0] == 0);
    REQUIRE(result[1] == 3);
    REQUIRE(result[2] == 5);
}

TEST_CASE("Parallel tasks with sufficient resources") {
    std::vector<int> duration = {2, 3, 2};
    std::vector<int> resource_requirement = {2, 2, 2};
    std::vector<std::vector<int>> dependencies = {{}, {}, {}};
    int resource_limit = 6;

    auto result = task_scheduler::schedule(3, duration, resource_requirement, dependencies, resource_limit);
    
    REQUIRE(result.size() == 3);
    REQUIRE(result[0] == 0);
    REQUIRE(result[1] == 0);
    REQUIRE(result[2] == 0);
}

TEST_CASE("Resource constraint prevents parallel execution") {
    std::vector<int> duration = {2, 3, 2};
    std::vector<int> resource_requirement = {3, 3, 3};
    std::vector<std::vector<int>> dependencies = {{}, {}, {}};
    int resource_limit = 3;

    auto result = task_scheduler::schedule(3, duration, resource_requirement, dependencies, resource_limit);
    
    REQUIRE(result.size() == 3);
    int makespan = 0;
    for(size_t i = 0; i < result.size(); i++) {
        makespan = std::max(makespan, result[i] + duration[i]);
    }
    REQUIRE(makespan == 7);
}

TEST_CASE("Cyclic dependencies") {
    std::vector<int> duration = {2, 2, 2};
    std::vector<int> resource_requirement = {1, 1, 1};
    std::vector<std::vector<int>> dependencies = {{1}, {2}, {0}};
    int resource_limit = 3;

    auto result = task_scheduler::schedule(3, duration, resource_requirement, dependencies, resource_limit);
    
    REQUIRE(result.empty());
}

TEST_CASE("Complex dependency chain") {
    std::vector<int> duration = {2, 3, 4, 2, 3};
    std::vector<int> resource_requirement = {2, 2, 2, 2, 2};
    std::vector<std::vector<int>> dependencies = {{}, {0}, {1}, {2}, {1, 3}};
    int resource_limit = 4;

    auto result = task_scheduler::schedule(5, duration, resource_requirement, dependencies, resource_limit);
    
    REQUIRE(result.size() == 5);
    for(size_t i = 0; i < dependencies.size(); i++) {
        for(int dep : dependencies[i]) {
            REQUIRE(result[i] >= result[dep] + duration[dep]);
        }
    }
}

TEST_CASE("Insufficient resources") {
    std::vector<int> duration = {2, 3, 2};
    std::vector<int> resource_requirement = {5, 5, 5};
    std::vector<std::vector<int>> dependencies = {{}, {}, {}};
    int resource_limit = 4;

    auto result = task_scheduler::schedule(3, duration, resource_requirement, dependencies, resource_limit);
    
    REQUIRE(result.empty());
}

TEST_CASE("Single task") {
    std::vector<int> duration = {5};
    std::vector<int> resource_requirement = {3};
    std::vector<std::vector<int>> dependencies = {{}};
    int resource_limit = 3;

    auto result = task_scheduler::schedule(1, duration, resource_requirement, dependencies, resource_limit);
    
    REQUIRE(result.size() == 1);
    REQUIRE(result[0] == 0);
}

TEST_CASE("Maximum size input") {
    std::vector<int> duration(100, 1);
    std::vector<int> resource_requirement(100, 1);
    std::vector<std::vector<int>> dependencies(100);
    for(int i = 1; i < 100; i++) {
        dependencies[i] = {i-1};
    }
    int resource_limit = 100;

    auto result = task_scheduler::schedule(100, duration, resource_requirement, dependencies, resource_limit);
    
    REQUIRE(result.size() == 100);
    for(int i = 1; i < 100; i++) {
        REQUIRE(result[i] > result[i-1]);
    }
}

TEST_CASE("Complex resource sharing") {
    std::vector<int> duration = {3, 2, 4, 3, 2};
    std::vector<int> resource_requirement = {3, 2, 2, 3, 1};
    std::vector<std::vector<int>> dependencies = {{}, {}, {0, 1}, {2}, {2}};
    int resource_limit = 5;

    auto result = task_scheduler::schedule(5, duration, resource_requirement, dependencies, resource_limit);
    
    REQUIRE(result.size() == 5);
    // Verify resource constraints
    std::vector<int> timeline(20, 0);  // Assuming makespan won't exceed 20
    for(size_t i = 0; i < result.size(); i++) {
        for(int t = result[i]; t < result[i] + duration[i]; t++) {
            timeline[t] += resource_requirement[i];
            REQUIRE(timeline[t] <= resource_limit);
        }
    }
}